Source/CombatExtended/CombatExtended/Graphic_Smoke.cs

```diff
@@ -0,0 +1,29 @@
+﻿using RimWorld;
+using UnityEngine;
+using Verse;
+
+namespace CombatExtended
+{
+    public class Graphic_Smoke : Graphic_Gas
+    {
+        private const float DistinctAlphaLevels = 128f;
+
+        public override void DrawWorker(Vector3 loc, Rot4 rot, ThingDef thingDef, Thing thing, float extraRotation)
+        {
+            Rand.PushState();
+            Rand.Seed = thing.thingIDNumber.GetHashCode();
+            Smoke smoke = thing as Smoke;
+            // round alpha to avoid creating too many new materials
+            var alpha = Mathf.Round(smoke.GetOpacity() * DistinctAlphaLevels) / DistinctAlphaLevels;
+            var materialColor = new Color(color.r, color.g, color.r, color.a * alpha);
+            var material = MaterialPool.MatFrom(new MaterialRequest((Texture2D)MatSingle.mainTexture, MatSingle.shader, materialColor));
```

Comment:
Do these ever get cleaned up or cached or do they just keep leaking memory?

Reply:
The materials, specifically? The pool caches them in a dictionary after creation, and uses the request as a key.